include/sys/fs/zfs.h

```diff
@@ -1282,16 +1286,18 @@ typedef enum zfs_ioc {
 	/*
 	 * Linux - 3/64 numbers reserved.
 	 */
-	ZFS_IOC_LINUX = ('Z' << 8) + 0x80,
+	ZFS_IOC_LINUX = ZFS_IOC_FIRST + 0x80,
 	ZFS_IOC_EVENTS_NEXT,			/* 0x5a81 */
 	ZFS_IOC_EVENTS_CLEAR,			/* 0x5a82 */
 	ZFS_IOC_EVENTS_SEEK,			/* 0x5a83 */
 
 	/*
 	 * FreeBSD - 1/64 numbers reserved.
```

Comment:
I honestly don't know either. OSX port been out since 2013 and only use one. Although, at some point we want to do zvol plumb/unplumb perhaps.  It is a hassle to move the range (with old binaries being an issue), when perhaps, the "uncleanliness" of having potentially two ranges, for say freebsd, (+0x80 and +0x100) or whatever. But I can probably move either range here if needed. Then it's the situation where Linux defined the three events one, that osx and windows also use. So they "could have" been in the generic pool.

Reply:
What if we instead abandoned the notion of reserving ranges per-platform.  We could declare `0x80` and up simply as optional ioctls which may, or may not, be implemented on a per-platform basis.  So for now it would be:

```
        /*
         * Per-platform (Optional) - 7/128 numbers reserved.
         */
        ZFS_IOC_PLATFORM = ZFS_IOC_FIRST + 0x80,
        ZFS_IOC_EVENTS_NEXT,                    /* 0x81 (Linux, MacOS, Windows) */
        ZFS_IOC_EVENTS_CLEAR,                   /* 0x82 (Linux, MacOS, Windows) */
        ZFS_IOC_EVENTS_SEEK,                    /* 0x83 (Linux, MacOS, Windows) */
        ZFS_IOC_NEXTBOOT,                       /* 0x84 (FreeBSD) */
	ZFS_IOC_JAIL,                           /* 0x85 (FreeBSD) */
	ZFS_IOC_UNJAIL,                         /* 0x86 (FreeBSD) */
        ZFS_IOC_PROXY_DATASET,                  /* 0x87 (MacOS) */
```

We could omit the MacOS and Windows comments until those platforms are included.  This would leave us the full range, and allow any platform to implement what they see fit as long as they reserve it first.

Reply:
In principle that is just fine by me, but if you were to move `ZFS_IOC_PROXY_DATASET` from 0xd0 to 0x87, then suddenly, should a user accidentally use old `zpool` userland, talking to a new kernel, there will be trouble. Removing the pre-assigned range as a concept is great, moving the number they already have, not so much :)

Reply:
Absolutely.  No need to cause any compatibility troubles if we can at all avoid it.  But can we?  if I'm not mistaken aren't we're already suffering from the renumbering problem for some of the common ioctl numbers between the platforms.  macos, freebsd, linux, and illumos all appear to differ slightly.

Reply:
To summarize I think we have 3 options here:

1) @mattmacy reminded me that ZoF included functionality to remap legacy ioctls to new ones which might be adaptable for macos.
2) While it would be nice to share one  `zfs_ioc_t` we could allow each platform to define their own.  This way each platform can retain their existing ordering without compatibly problems.  This might be the best solution for windows anyway.
3) Cut over cold turkey, and make the old utilities refuse to interoperate with the new kernel modules.  Not a good user experience, but an option.

@lundman I'm OK with however you'd like to handle this on macos.  I just want to make sure whatever we decide here for Linux and FreeBSD is going to be acceptable for you.

Reply:
It is possible we have re-ordering already, I had not yet looked, since a platform only needs to be internally consistent. Looks like OSX/ZOL are identical until `ZFS_IOC_DESTROY_BOOKMARKS` then we diverge wildly. 

How would 1) work? Not sure we should keep a legacy map forever, but for "one or so" releases it could be nice. Since we got `zpool version` already from ZOL, we could more easily do a hard cut over, and warn users about it.

Reply:
@lundman I'm obviously fine with OSX doing whatever you like. However, I've gone through the sunk cost to consolidate. FreeBSD passes in a pointer to `zfs_iocparm_t` which as a version field. This has allowed it to version across a bunch of changes. `zfs_cmd_t` is also different in HEAD so I have to maintain legacy interface for the old tools to work on the new kmod. If you have the same `zfs_cmd_t` structure you just need to remap the ioctl number. This is what it looks like on FreeBSD:

```
unsigned static long zfs_ioctl_bsd12_to_zof[] = {
	ZFS_IOC_POOL_CREATE,			/* 0x00 */
	ZFS_IOC_POOL_DESTROY,			/* 0x01 */
	ZFS_IOC_POOL_IMPORT,			/* 0x02 */
	ZFS_IOC_POOL_EXPORT,			/* 0x03 */
	ZFS_IOC_POOL_CONFIGS,			/* 0x04 */
	ZFS_IOC_POOL_STATS,			/* 0x05 */
	ZFS_IOC_POOL_TRYIMPORT,			/* 0x06 */
	ZFS_IOC_POOL_SCAN,			/* 0x07 */
	ZFS_IOC_POOL_FREEZE,			/* 0x08 */
	ZFS_IOC_POOL_UPGRADE,			/* 0x09 */
	ZFS_IOC_POOL_GET_HISTORY,		/* 0x0a */
	ZFS_IOC_VDEV_ADD,			/* 0x0b */
	ZFS_IOC_VDEV_REMOVE,			/* 0x0c */
	ZFS_IOC_VDEV_SET_STATE,			/* 0x0d */
	ZFS_IOC_VDEV_ATTACH,			/* 0x0e */
	ZFS_IOC_VDEV_DETACH,			/* 0x0f */
	ZFS_IOC_VDEV_SETPATH,			/* 0x10 */
	ZFS_IOC_VDEV_SETFRU,			/* 0x11 */
	ZFS_IOC_OBJSET_STATS,			/* 0x12 */
	ZFS_IOC_OBJSET_ZPLPROPS,		/* 0x13 */
	ZFS_IOC_DATASET_LIST_NEXT,		/* 0x14 */
	ZFS_IOC_SNAPSHOT_LIST_NEXT,		/* 0x15 */
	ZFS_IOC_SET_PROP,			/* 0x16 */
	ZFS_IOC_CREATE,				/* 0x17 */
	ZFS_IOC_DESTROY,			/* 0x18 */
	ZFS_IOC_ROLLBACK,			/* 0x19 */
	ZFS_IOC_RENAME,				/* 0x1a */
	ZFS_IOC_RECV,				/* 0x1b */
	ZFS_IOC_SEND,				/* 0x1c */
	ZFS_IOC_INJECT_FAULT,			/* 0x1d */
	ZFS_IOC_CLEAR_FAULT,			/* 0x1e */
	ZFS_IOC_INJECT_LIST_NEXT,		/* 0x1f */
	ZFS_IOC_ERROR_LOG,			/* 0x20 */
	ZFS_IOC_CLEAR,				/* 0x21 */
	ZFS_IOC_PROMOTE,			/* 0x22 */
	/* start of mismatch */
	ZFS_IOC_DESTROY_SNAPS,			/* 0x23:0x3b */
	ZFS_IOC_SNAPSHOT,			/* 0x24:0x23 */
	ZFS_IOC_DSOBJ_TO_DSNAME,		/* 0x25:0x24 */
	ZFS_IOC_OBJ_TO_PATH,			/* 0x26:0x25 */
	ZFS_IOC_POOL_SET_PROPS,			/* 0x27:0x26 */
	ZFS_IOC_POOL_GET_PROPS,			/* 0x28:0x27 */
	ZFS_IOC_SET_FSACL,			/* 0x29:0x28 */
	ZFS_IOC_GET_FSACL,			/* 0x30:0x29 */
	ZFS_IOC_SHARE,				/* 0x2b:0x2a */
	ZFS_IOC_INHERIT_PROP,			/* 0x2c:0x2b */
	ZFS_IOC_SMB_ACL,			/* 0x2d:0x2c */
	ZFS_IOC_USERSPACE_ONE,			/* 0x2e:0x2d */
	ZFS_IOC_USERSPACE_MANY,			/* 0x2f:0x2e */
	ZFS_IOC_USERSPACE_UPGRADE,		/* 0x30:0x2f */
	ZFS_IOC_HOLD,				/* 0x31:0x30 */
	ZFS_IOC_RELEASE,			/* 0x32:0x31 */
	ZFS_IOC_GET_HOLDS,			/* 0x33:0x32 */
	ZFS_IOC_OBJSET_RECVD_PROPS,		/* 0x34:0x33 */
	ZFS_IOC_VDEV_SPLIT,			/* 0x35:0x34 */
	ZFS_IOC_NEXT_OBJ,			/* 0x36:0x35 */
	ZFS_IOC_DIFF,				/* 0x37:0x36 */
	ZFS_IOC_TMP_SNAPSHOT,			/* 0x38:0x37 */
	ZFS_IOC_OBJ_TO_STATS,			/* 0x39:0x38 */
	ZFS_IOC_JAIL,			/* 0x3a:0xc2 */
	ZFS_IOC_UNJAIL,			/* 0x3b:0xc3 */
	ZFS_IOC_POOL_REGUID,			/* 0x3c:0x3c */
	ZFS_IOC_SPACE_WRITTEN,			/* 0x3d:0x39 */
	ZFS_IOC_SPACE_SNAPS,			/* 0x3e:0x3a */
	ZFS_IOC_SEND_PROGRESS,			/* 0x3f:0x3e */
	ZFS_IOC_POOL_REOPEN,			/* 0x40:0x3d */
	ZFS_IOC_LOG_HISTORY,			/* 0x41:0x3f */
	ZFS_IOC_SEND_NEW,			/* 0x42:0x40 */
	ZFS_IOC_SEND_SPACE,			/* 0x43:0x41 */
	ZFS_IOC_CLONE,				/* 0x44:0x42 */
	ZFS_IOC_BOOKMARK,			/* 0x45:0x43 */
	ZFS_IOC_GET_BOOKMARKS,			/* 0x46:0x44 */
	ZFS_IOC_DESTROY_BOOKMARKS,		/* 0x47:0x45 */
	ZFS_IOC_NEXTBOOT,			/* 0x48:0xc1 */
	ZFS_IOC_CHANNEL_PROGRAM,		/* 0x49:0x48 */
	ZFS_IOC_REMAP,				/* 0x4a:0x4c */
	ZFS_IOC_POOL_CHECKPOINT,		/* 0x4b:0x4d */
	ZFS_IOC_POOL_DISCARD_CHECKPOINT,	/* 0x4c:0x4e */
	ZFS_IOC_POOL_INITIALIZE,		/* 0x4d:0x4f */
};
```
The ioctl code is as follows:

```
static int
zfsdev_ioctl(struct cdev *dev, u_long zcmd, caddr_t arg, int flag,
    struct thread *td)
{
	uint_t len, vecnum;
	zfs_iocparm_t *zp;
	zfs_cmd_t *zc;
	zfs_cmd_legacy_t *zcl;
	int rc, error;
	void *uaddr;

	len = IOCPARM_LEN(zcmd);
	vecnum = zcmd & 0xff;
	zp = (void *)arg;
	uaddr = (void *)zp->zfs_cmd;
	error = 0;
	zcl = NULL;

	if (len != sizeof (zfs_iocparm_t)) {
		return (EINVAL);
	}

	zc = kmem_zalloc(sizeof (zfs_cmd_t), KM_SLEEP);
	/*
	 * Remap ioctl code for legacy user binaries
	 */
	if (zp->zfs_ioctl_version == ZFS_IOCVER_FREEBSD) {
		if (vecnum >= sizeof (zfs_ioctl_bsd12_to_zof)/sizeof (long)) {
			kmem_free(zc, sizeof (zfs_cmd_t));
			return (ENOTSUP);
		}
		zcl = kmem_zalloc(sizeof (zfs_cmd_legacy_t), KM_SLEEP);
		vecnum = zfs_ioctl_bsd12_to_zof[vecnum];
		if (copyin(uaddr, zcl, sizeof(zfs_cmd_legacy_t))) {
			error = SET_ERROR(EFAULT);
			goto out;
		}
		zfs_cmd_bsd12_to_zof(zcl, zc);
	} else if (copyin(uaddr, zc, sizeof(zfs_cmd_t))) {
		error = SET_ERROR(EFAULT);
		goto out;
	}
	error = zfsdev_ioctl_common(vecnum, zc);
	if (zcl) {
		zfs_cmd_zof_to_bsd12(zc, zcl);
		rc = copyout(zcl, uaddr, sizeof(*zcl));
	} else {
		rc = copyout(zc, uaddr, sizeof(*zc));
	}
	if (error == 0 && rc != 0)
		error = SET_ERROR(EFAULT);
 out:
	if (zcl)
		kmem_free(zcl, sizeof (zfs_cmd_legacy_t));
	kmem_free(zc, sizeof (zfs_cmd_t));
	return (error);
}```

This is why there is now a `zfs_ioctl_fd` routine wherever zfs ioctls are used:
```
tatic int
zcmd_ioctl_compat(int fd, int request, zfs_cmd_t *zc, const int cflag)
{
	int ret;
	void *zc_c;
	unsigned long ncmd;
	zfs_iocparm_t zp;

	switch (cflag) {
	case ZFS_CMD_COMPAT_NONE:
		ncmd = _IOWR('Z', request, zfs_iocparm_t);
		zp.zfs_cmd = (uint64_t)zc;
		zp.zfs_cmd_size = sizeof (zfs_cmd_t);
		zp.zfs_ioctl_version = ZFS_IOCVER_ZOF;
		return (ioctl(fd, ncmd, &zp));
	default:
		abort();
		return (EINVAL);
	}

	ret = ioctl(fd, ncmd, zc_c);
	zfs_cmd_compat_get(zc, (caddr_t)zc_c, cflag);
	free(zc_c);

	return (ret);
}

/*
 * This is FreeBSD version of ioctl, because Solaris' ioctl() updates
 * zc_nvlist_dst_size even if an error is returned, on FreeBSD if an
 * error is returned zc_nvlist_dst_size won't be updated.
 */
int
zfs_ioctl_fd(int fd, unsigned long request, zfs_cmd_t *zc)
{
	size_t oldsize;
	int ret, cflag = ZFS_CMD_COMPAT_NONE;

	oldsize = zc->zc_nvlist_dst_size;
	ret = zcmd_ioctl_compat(fd, request, zc, cflag);

	if (ret == 0 && oldsize < zc->zc_nvlist_dst_size) {
		ret = -1;
		errno = ENOMEM;
	}

	return (ret);
}
```